### ISSUE TITLE
figure out why k-space drawing for the sinc interpolant is so slow

### ISSUE THREAD

**Issue by beckermr** [GalSim-developers/JAX-GalSim#81]
The k-space drawing for the sinc interpolant is really slow. We should figure out why and try to fix it.

**Comment by AdamField118** [comment]
So looking into this I ran the diagnoses:

```
galsim      quintic k_interp  ok     0.00s  (9, 9)
galsim      SINC    k_interp  ok     0.10s  (9, 9)
jax_galsim  quintic k_interp  ok     2.70s  (9, 9)
jax_galsim  SINC    k_interp  FAIL   0.25s  Out of memory allocating 5252700183164 bytes.
```

(Ignore line three since the `jax_galsim quintic` time is jit-compile dominated in this; see line 4 for the issue)

`_draw_with_interpolant_kval`, [linked here](https://github.com/GalSim-developers/JAX-GalSim/blob/e44e1a9d9857b42166d88aca4b2e6ef770875657/jax_galsim/interpolatedimage.py#L1134-L1135), sizes its stencil from `ixrange` and forms a [dense 2D outer product](https://github.com/GalSim-developers/JAX-GalSim/blob/e44e1a9d9857b42166d88aca4b2e6ef770875657/jax_galsim/interpolatedimage.py#L1155-L1158) over it:

```python
irange = interp.ixrange // 2
iinds = jnp.arange(-irange, irange + 1)
...
wkx = wkx[None, :, :];  wky = wky[:, None, :]
```

That is $(\texttt{ixrange}+1)^2 \cdot \texttt{npix}$ elements. For sinc, $\texttt{xrange} = 1/(\pi\cdot\texttt{kvalue}\textunderscore\texttt{accuracy})$ ([linked here](https://github.com/GalSim-developers/JAX-GalSim/blob/e44e1a9d9857b42166d88aca4b2e6ef770875657/jax_galsim/interpolant.py#L431)) gives `ixrange = 63662` at the default `kvalue_accuracy=1e-5`, against `6` for both `Quintic` and `Lanczos(3)`. Sweeping `kvalue_accuracy` confirms the scaling, the request tracks $2 \cdot (\texttt{ixrange}+1)^2 \cdot \texttt{npix} \cdot \texttt{itemsize}$ exactly:

```
 kvalue_acc  ixrange  result     observed bytes   (ix+1)^2*npix elems  bytes/elem
       0.01       64      ok      -- ran (9, 9)               342,225
      0.001      638      ok      -- ran (9, 9)            33,074,001
      1e-05    63662     OOM  5,252,700,183,164       328,291,183,089       16.00
                                                              (complex64)

# same sweep under JAX_ENABLE_X64=1:
      1e-05    63662     OOM 10,505,400,366,200       328,291,183,089       32.00
                                                             (complex128)
```

GalSim uses the same truncated $\pm\texttt{xrange}$ stencil, but evaluates it as a clipped, wrapped loop rather than materializing it, with a shortcut that [collapses the sum to one element on grid points](https://github.com/GalSim-developers/GalSim/blob/701524c7cd2d55aaa8036675c2d9420ff4a7a77e/src/SBInterpolatedImage.cpp#L412-L424).

So I was thinking about it, the k-space interpolation is already periodic, [`_interp_weight_1d_kval`](https://github.com/GalSim-developers/JAX-GalSim/blob/e44e1a9d9857b42166d88aca4b2e6ef770875657/jax_galsim/interpolatedimage.py#L1098-L1105) wraps every index (`kxind = (kxi + ioff) % nkx`) while weighting with the aperiodic sinc. Summing the aperiodic sinc over a periodic grid is [Whittaker–Shannon band-limited interpolation](https://en.wikipedia.org/wiki/Whittaker%E2%80%93Shannon_interpolation_formula) of one period, and that sum has a closed form: the Dirichlet kernel, which GalSim already carries as `SincInterpolant::xvalWrapped`, [linked here](https://github.com/GalSim-developers/GalSim/blob/701524c7cd2d55aaa8036675c2d9420ff4a7a77e/src/Interpolant.cpp#L213-L224).

```cpp
x *= M_PI;
if (N%2==0) return std::sin(x) * std::cos(x/N) / (N*std::sin(x/N));
else        return std::sin(x) / (N*std::sin(x/N));
```

So the stencil should span exactly one period (`nkx`, `nky`), not `ixrange`.

It matches the reference implementation too. It is not an approximation of the current behaviour, it removes an approximation. Interpolating a random complex signal on a periodic grid, closed form vs the aperiodic sum truncated at $\pm M$:

```
    N        x   |truncated - closed|   note
    8    0.000              5.333e-12   on grid -> exact
    8    3.500              3.019e-06   off grid -> trunc. error
    9    0.000              1.717e-15   on grid -> exact
    9    2.317              4.712e-06   off grid -> trunc. error
   32    1.000              8.646e-13   on grid -> exact
   32    2.317              8.129e-06   off grid -> trunc. error
  128    0.000              2.268e-12   on grid -> exact
  128    2.317              1.692e-05   off grid -> trunc. error
```

On grid points the two agree to machine precision. Off grid they differ by the truncation error of the current stencil, and that error shrinks toward the closed form as the window grows:

```
closed form, 32 terms = 0.685440112770+0.516110053942j

  M (half-width)      terms   |truncated - closed|
            3183       6367              4.103e-05
           31831      63663              1.157e-05   <- galsim default
          318310     636621              2.029e-06
         3183098    6366197              6.910e-08
```

At GalSim's default the current approach spends $63,663$ terms to land $\sim 1e-5$ away from an answer the closed form gives exactly in $N$ terms. So against reference GalSim we would agree to machine precision on grid points and differ by $\sim 1e-5$ off grid, with the closed form being the correct value of the two. This is a deliberate divergence from bit-matching GalSim, and a point I'd want to hear from you on.

This change saves:

| | stencil per axis | elements, 9×9 draw |
|---|---|---|
| today (truncated sinc) | 63,663 | 3.28e11 |
| Dirichlet closed form | 128 | 1.33e6 |

$\sim 2.5 \cdot 10^5$ fewer elements here, and the stencil stops depending on `kvalue_accuracy` altogether, it becomes a property of the k-image size, so the pathological case disappears rather than being pushed further out.

**Comment by beckermr** [comment]
If there is a more accurate and efficient way to do things, then we should implement it for sure. Also it sounds like we should make a pr to galsim itself with the better implementation.
